### PR TITLE
LG-12330 update design system

### PR DESCRIPTION
**Pull request by Sgtpluck**
### Relevant Ticket or Conversation:
https://cm-jira.usa.gov/browse/LG-12330

### Description of Changes:
This is the first step of https://cm-jira.usa.gov/browse/LG-12330. Bumping the design system to 3.0 resulted in segfaults when starting the server. It became necessary to migrate away from ruby-sass (which reached end of life March 26, 2019). I chose to utilize Propshaft for the asset pipeline to make the application similar to the IdP, and the changes just became fairly extensive, so it seemed like an appropriate time to get a review!

### PR Checklist:

1. [ ] Have you linted and tested your code locally prior to submission?
2. [ ] Have you tagged the appropriate dev(s) for review?
3. [ ] Have you linked to any relevant tickets or conversations?

### PR Review Standards: 
- Consider using [Conventional Comments](https://conventionalcomments.org/) to ensure that your feedback is clear and actionable.
- Ideally, PRs should be reviewed by at least 2 team members.
- All PRs must be approved before being merged.

### CODE REVIEW

File: config/initializers/assets.rb [hunk 1]
```diff
@@ -1,39 +1,11 @@
-# Be sure to restart your server when you modify this file.
+# require 'extensions/propshaft/asset'
```
Comment (aduth) [inline config/initializers/assets.rb:1]: Was this for debugging?

```suggestion
```

File: package.json [hunk 2]
```diff
@@ -1,24 +1,27 @@
...
   "license": "CC0-1.0",
   "engines": {
-    "node": ">=12",
+    "node": ">=18",
     "npm": "6.x.x"
   },
   "scripts": {
     "lint": "eslint . --ignore-pattern public --ignore-pattern vendor --ignore-pattern app/assets/builds",
-    "build": "webpack"
+    "build": "webpack --config webpack.config.js",
```
Comment (aduth) [inline package.json:9]: I would assume the `--config` value here would be default anyways and not necessary to specify?

```suggestion
    "build": "webpack",
```

File: package.json [hunk 3]
```diff
@@ -1,24 +1,27 @@
...
     "@babel/preset-env": "^7.23.2",
     "@rails/ujs": "^6.1.6",
     "babel-loader": "^9.1.3",
     "babel-plugin-polyfill-corejs3": "^0.8.0",
     "core-js": "^3.23.4",
     "fast-glob": "^3.2.11",
-    "identity-style-guide": "^2.2.3",
-    "webpack": "^5.76.0",
-    "webpack-cli": "^4.10.0"
+    "identity-style-guide": "3.0.0",
```
Comment (aduth) [inline package.json:21]: 3.0.0 was a pretty big jump for USWDS (from 2.0.3 to 2.9.0) and had a lot of breaking changes. Have you made a pass over those? For example, I'd expect to see some revisions to the base layout for banner, etc (SVG images, banner markup).

File: package.json [hunk 4]
```diff
@@ -1,24 +1,27 @@
...
+    "sass": "^1.71.1",
```
Comment (aduth) [inline package.json:22]: I wouldn't expect `sass` should need to be installed. `@18f/identity-build-sass` internally uses `sass-embedded`, which is a different implementation of the Sass compiler.

File: bin/dev [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env sh
+
+if gem list --no-installed --exact --silent foreman; then
+  echo "Installing foreman..."
+  gem install foreman
+fi
```
Comment (aduth) [inline bin/dev:6]: Wouldn't this be expected to be installed through the `make setup` process?

https://github.com/18F/identity-dashboard/blob/52efcfcae6ff115355c0d64b8779b9a56a08e6d6/bin/setup#L44

File: bin/dev [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+# Default to port 3000 if not specified
```
Comment (aduth) [inline bin/dev:8]: Typo: The default is 3001

```suggestion
# Default to port 3001 if not specified
```

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -177,15 +176,15 @@
...
             </div>
 
           </div>
         </div>
       </div>
     </footer>
 
-    <%= javascript_include_tag 'js/main.js' %>
-    <%= javascript_include_tag "application" %>
+    <%= javascript_include_tag asset_url('main.js') %>
```
Comment (aduth) [inline app/views/layouts/application.html.erb:187]: I wouldn't expect you'd have to use `asset_url` in any of the code changes, was it necessary?

File: app/assets/stylesheets/application.sass.scss
Comment (aduth) [inline app/assets/stylesheets/application.sass.scss:1]: Is this file necessary?

File: app/javascript/application.js
Comment (aduth) [inline app/javascript/application.js:1]: Same question about whether it's necessary

File: bin/dev [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+export PORT="${PORT:-3001}"
+
+exec foreman start -f Procfile "$@"
```
Comment (aduth) [inline bin/dev:11]: If there's already a default specified in `Procfile`, does it need to be re-specified here?

I'd wonder if we need this file at all, or if the original Makefile target code would suffice (`foreman start`)

File: config/initializers/assets.rb (hunk 1 above)
Comment (Sgtpluck) [reply config/initializers/assets.rb:1]: oh that was copied from the idp and i forgot to remove it! thanks

File: app/assets/stylesheets/application.sass.scss
Comment (Sgtpluck) [reply app/assets/stylesheets/application.sass.scss:1]: i will try removing it and see!

File: app/javascript/application.js
Comment (Sgtpluck) [reply app/javascript/application.js:1]: i'll remove it!

File: app/views/layouts/application.html.erb (hunk 7 above)
Comment (Sgtpluck) [reply app/views/layouts/application.html.erb:187]: it definitely wasn't initially working without it, but a lot of other things have changed, so maybe it's possible now. i'll see!

File: config/initializers/assets.rb (hunk 1 above)
Comment (Sgtpluck) [reply config/initializers/assets.rb]: updated in https://github.com/18F/identity-dashboard/pull/740/commits/594bc2dafc173305e4c2b377dcf9ad97baabc8e9

File: package.json (hunk 2 above)
Comment (Sgtpluck) [reply package.json]: updated in https://github.com/18F/identity-dashboard/commit/594bc2dafc173305e4c2b377dcf9ad97baabc8e9

File: package.json (hunk 4 above)
Comment (Sgtpluck) [reply package.json]: updated in https://github.com/18F/identity-dashboard/commit/594bc2dafc173305e4c2b377dcf9ad97baabc8e9

File: app/views/layouts/application.html.erb (hunk 7 above)
Comment (Sgtpluck) [reply app/views/layouts/application.html.erb]: updated in https://github.com/18F/identity-dashboard/commit/594bc2dafc173305e4c2b377dcf9ad97baabc8e9 seems to work!

File: app/assets/stylesheets/application.sass.scss
Comment (Sgtpluck) [reply app/assets/stylesheets/application.sass.scss:1]: updated in https://github.com/18F/identity-dashboard/commit/594bc2dafc173305e4c2b377dcf9ad97baabc8e9

File: app/javascript/application.js
Comment (Sgtpluck) [reply app/javascript/application.js:1]: updated in https://github.com/18F/identity-dashboard/commit/594bc2dafc173305e4c2b377dcf9ad97baabc8e9

File: bin/dev (hunk 8 above)
Comment (Sgtpluck) [reply bin/dev]: updated in https://github.com/18F/identity-dashboard/commit/594bc2dafc173305e4c2b377dcf9ad97baabc8e9

File: package.json (hunk 3 above)
Comment (Sgtpluck) [reply package.json:21]: I walked through the release notes from 2.0.3 to 2.9 and I think I got most if not everything!

**Review by mitchellhenke — APPROVED** [review]

File: app/assets/stylesheets/application.css.scss [hunk 9]
```diff
@@ -77,22 +75,16 @@ $image-path: 'img';
...
-  background: red;
+  background: $theme-color-secondary;
+  color: $theme-color-base-lightest;
 }
 
 .certificate-warning,
 .usa-table td.certificate-warning {
-  color: black;
-  background: yellow;
+  background: $theme-color-warning;
```
Comment (aduth) [inline app/assets/stylesheets/application.css.scss:86]: Nice updates to use the standard colors 👍

File: app/assets/stylesheets/application.css.scss [hunk 10]
```diff
@@ -127,23 +119,11 @@ $image-path: 'img';
...
-
-.usa-search-input--joined-right {
-  border-top-right-radius: 0;
-  border-bottom-right-radius: 0;
-  max-width: 12rem;
-}
-
-.text-heavy {
-  font-weight: 800;
-}
```
Comment (aduth) [inline app/assets/stylesheets/application.css.scss:149]: I see a few references to this class. Do those need to be updated? I could imagine `text-bold` could work as a close, standard alternative.

https://github.com/18F/identity-dashboard/blob/1a6a438a8be5f442e478786eaca2ec0bd5480707/app/views/tools/saml_request.html.erb#L45-L55

https://github.com/18F/identity-dashboard/blob/1a6a438a8be5f442e478786eaca2ec0bd5480707/config/locales/service_providers.en.yml#L4-L5

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -37,54 +36,69 @@
...
+                  <p>
+                    <strong>Secure .gov websites use HTTPS</strong>
+                    <br>
+                    A <strong>lock</strong> (
+                      <%= image_tag 'lock.svg',
+                        alt: 'A locked padlock',
+                        class: 'usa-banner__lock-image',
+                        role: 'img' %>
+                    ) or <strong>https://</strong> means you’ve safely connected to the .gov website. Share sensitive information only on official, secure websites.
+                    The <strong>https://</strong> ensures that you are connecting to the official website and that any information you provide is encrypted and transmitted securely.
```
Comment (aduth) [inline app/views/layouts/application.html.erb:87]: I think this second sentence should be removed? I think it's what's being updated.

```suggestion
```

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -37,54 +36,69 @@
...
-              <div class="usa-media-block__body">
-                <p>
-                  <strong>The .gov means it’s official.</strong>
-                  <br>
-                  Federal government websites often end in .gov or .mil. Before sharing sensitive information, make sure you’re on a federal government site.
-                </p>
+          <div class="usa-banner__content usa-accordion__content" id="gov-banner">
+            <%= javascript_tag(nonce: true) do %>
+              document.getElementById('gov-banner').setAttribute('hidden', '');
+            <% end %>
```
Comment (aduth) [inline app/views/layouts/application.html.erb:59]: I'm seeing a CSP error in my console when testing in local development.

For this to work, it may be necessary to configure nonce directives to at least include nonce in `script-src`:

https://github.com/18F/identity-dashboard/blob/1a6a438a8be5f442e478786eaca2ec0bd5480707/config/initializers/content_security_policy.rb#L21

See equivalent IdP code: https://github.com/18F/identity-idp/blob/0621238/config/initializers/content_security_policy.rb#L50

File: app/views/service_providers/_form.html.erb [hunk 13]
```diff
@@ -52,18 +52,18 @@
     <%= form.label :identity_protocol,
                    label: 'Authentication protocol',
                    class: 'usa-label' %>
-    <%= content_tag(:ul, class: 'usa-input-list') do
-      ServiceProvider.identity_protocols.each_key do |protocol| %>
-        <%= content_tag(:li) do %>
+    <div>
```
Comment (aduth) [inline app/views/service_providers/_form.html.erb:55]: Is the wrapping `<div>` necessary?

File: app/views/service_providers/_form.html.erb [hunk 14]
```diff
@@ -133,12 +131,12 @@
       <%= form.label :allow_prompt_login,
                       label: 'Enable <prompt=login>',
                       class: 'usa-label' %>
-      <ul class="usa-input-list">
-        <li>
+      <div>
```
Comment (aduth) [inline app/views/service_providers/_form.html.erb:134]: Same question about `<div>` wrapper. There's another example below too.

File: app/assets/stylesheets/application.css.scss (hunk 10 above)
Comment (Sgtpluck) [reply app/assets/stylesheets/application.css.scss:149]: `text-heavy` is a [uswds utility class](https://designsystem.digital.gov/utilities/text-styles/)! i double checked the spots it's used to make sure it was available in the version that's in this version of the design system, so i think it's ok.

File: app/views/service_providers/_form.html.erb (hunk 13 above)
Comment (Sgtpluck) [reply app/views/service_providers/_form.html.erb:55]: The [changelog](https://github.com/uswds/uswds/releases/tag/v2.0.0-beta.7) seemed to indicate replacing the `<ul>` with a `<div>` but i may have read it incorrectly.

File: app/views/layouts/application.html.erb (hunk 12 above)
Comment (Sgtpluck) [reply app/views/layouts/application.html.erb:59]: maybe i'll just take this out and worry about it another day; i grabbed it from the idp since i was using it as reference for the banner. don't really need to add csp scope to this pr

File: app/views/service_providers/_form.html.erb (hunk 13 above)
Comment (Sgtpluck) [reply app/views/service_providers/_form.html.erb]: updated in https://github.com/18F/identity-dashboard/pull/740/commits/20e94541458f876cf2ab980a6a9c511e0bce11dc

File: app/views/service_providers/_form.html.erb (hunk 14 above)
Comment (Sgtpluck) [reply app/views/service_providers/_form.html.erb]: updated in https://github.com/18F/identity-dashboard/pull/740/commits/20e94541458f876cf2ab980a6a9c511e0bce11dc

File: app/views/layouts/application.html.erb (hunk 12 above)
Comment (Sgtpluck) [reply app/views/layouts/application.html.erb]: removed in https://github.com/18F/identity-dashboard/commit/20e94541458f876cf2ab980a6a9c511e0bce11dc

Comment (aduth) [reply app/views/layouts/application.html.erb]: Oh I see you added back the `hidden` attribute too, so that should solve that flickering.

File: .circleci/config.yml [hunk 15]
```diff
@@ -52,7 +52,7 @@ jobs:
           command: |
             gem install bundler
             bundle install --deployment --jobs=4 --retry=3 --without deploy development production --path vendor/bundle
-            npm install
+            yarn install --frozen-lockfile --ignore-engines --cache-folder .yarn-cache
```
[the remark this answers is not in the document]
Comment (Sgtpluck) [reply .circleci/config.yml:55]: 👍 i can add it in separately -- i'd like to be able to get this PR merged in so i can moved forward with the rest of the design system bumps!

**Review by aduth — APPROVED** [review]
LGTM ! 🎉
